File: dlox/lib/src/interpreter.dart
import 'builtins.dart';
import 'callable.dart';
import 'class.dart';
import 'environment.dart';
import 'exception.dart';
import 'expression.dart';
import 'function.dart';
import 'instance.dart';
import 'lox.dart';
import 'nil.dart';
import 'return.dart';
import 'statement.dart';
import 'token.dart';
import 'util.dart';
import 'visitor.dart';

class Interpreter implements ExprVisitor<Object>, StmtVisitor<void> {
  final globals = Environment();
  final _locals = <Expr, int>{};
  var _env;

  Interpreter() {
    globals.define('clock', Clock());
    _env = globals;
  }

  void interpret(List<Stmt> statements) {
    try {
      statements.forEach(_execute);
    } on RuntimeError catch (e) {
      Lox.runtimeError(e);
    }
  }

  void resolve(Expr expr, int depth) => _locals[expr] = depth;

  @override
  Object visitAssignExpr(AssignExpr expr) {
    final value = _evaluate(expr.value);
    final distance = _locals[expr];

    if (distance != null) {
      _env.assignAt(distance, expr.name, value);
    } else {
      globals.assign(expr.name, value);
    }

    return value;
  }

  @override
  void visitBlockStmt(BlockStmt stmt) =>
      executeBlock(stmt.statements, Environment(_env));

  @override
  Object visitBinaryExpr(BinaryExpr expr) {
    final left = _evaluate(expr.left);
    final right = _evaluate(expr.right);

    switch (expr.operator.type) {
      case TokenType.MINUS:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) - (right as double);
      case TokenType.SLASH:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) / (right as double);
      case TokenType.STAR:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) * (right as double);
      case TokenType.PLUS:
        if (left is double && right is double) {
          return left + right;
        }

        if (left is String && right is String) {
          return left + right;
        }

        throw RuntimeError(
          expr.operator,
          'Operands must be two numbers or two strings.',
        );
      case TokenType.GREATER:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) > (right as double);
      case TokenType.GREATER_EQUAL:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) >= (right as double);
      case TokenType.LESS:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) < (right as double);
      case TokenType.LESS_EQUAL:
        _checkNumberOperands(expr.operator, left, right);
        return (left as double) <= (right as double);
      case TokenType.BANG_EQUAL:
        return !isEqual(left, right);
      case TokenType.EQUAL_EQUAL:
        return isEqual(left, right);
      default:
        throw UnimplementedError();
    }
  }

  @override
  Object visitCallExpr(CallExpr expr) {
    final callee = _evaluate(expr.callee);

    if (callee is! LoxCallable) {
      throw RuntimeError(expr.paren, 'Can only call functions and classes.');
    }

    final args = expr.arguments.map(_evaluate).toList();

    if (args.length != callee.arity) {
      throw RuntimeError(expr.paren,
          'Expected ${callee.arity} arguments but got ${args.length}.');
    }

    return callee.call(this, args);
  }

  @override
  void visitClassStmt(ClassStmt stmt) {
    var superclass;
    if (stmt.superclass != null) {
      superclass = _evaluate(stmt.superclass!);
      if (superclass is! LoxClass) {
        throw RuntimeError(
            stmt.superclass!.name, 'Superclass must be a class.');
      }
    }

    _env.define(stmt.name.lexeme, Nil());

    if (stmt.superclass != null) {
      _env = Environment(_env);
      _env.define('super', superclass);
    }

    final methods = stmt.methods.fold(
      <String, LoxFunction>{},
      (Map<String, LoxFunction> acc, method) {
        final lexeme = method.name.lexeme;
        acc[lexeme] = LoxFunction(method, _env, lexeme == 'init');
        return acc;
      },
    );

    final klass = LoxClass(stmt.name.lexeme, superclass, methods);

    if (superclass != null) {
      _env = _env.enclosing;
    }

    _env.assign(stmt.name, klass);
  }

  @override
  void visitExpressionStmt(ExpressionStmt stmt) => _evaluate(stmt.expression);

  @override
  void visitFunctionStmt(FunctionStmt stmt) =>
      _env.define(stmt.name.lexeme, LoxFunction(stmt, _env, false));

  @override
  Object visitGetExpr(GetExpr expr) {
    final obj = _evaluate(expr.object);
    if (obj is LoxInstance) {
      return obj.get(expr.name);
    }
    throw RuntimeError(expr.name, 'Only instances have properties.');
  }

  @override
  Object visitGroupingExpr(GroupingExpr expr) => _evaluate(expr.expression);

  @override
  void visitIfStmt(IfStmt stmt) {
    if (isTruthy(_evaluate(stmt.condition))) {
      _execute(stmt.thenBranch);
    } else if (stmt.elseBranch != null) {
      _execute(stmt.elseBranch!);
    }
  }

  @override
  Object visitLiteralExpr(LiteralExpr expr) => expr.value;

  @override
  Object visitLogicalExpr(LogicalExpr expr) {
    final left = _evaluate(expr.left);

    if (expr.operator.type == TokenType.OR) {
      if (isTruthy(left)) return left;
    } else {
      if (!isTruthy(left)) return left;
    }

    return _evaluate(expr.right);
  }

  @override
  Object visitNil(Nil expr) => expr;

  @override
  void visitPrintStmt(PrintStmt stmt) =>
      print(stringify(_evaluate(stmt.expression)));

  @override
  void visitReturnStmt(ReturnStmt stmt) {
    var value;

    if (stmt.value is! Nil) {
      value = _evaluate(stmt.value);
    }

    value ??= Nil();

    throw Return(value);
  }

  @override
  Object visitSetExpr(SetExpr expr) {
    final obj = _evaluate(expr.object);

    if (obj is! LoxInstance) {
      throw RuntimeError(expr.name, 'Only instances have fields.');
    }

    final value = _evaluate(expr.value);
    obj.set(expr.name, value);
    return value;
  }

  @override
  Object visitSuperExpr(SuperExpr expr) {
    final distance = _locals[expr]!;
    final superclass = _env.getAt(distance, 'super');
    final obj = _env.getAt(distance - 1, 'this');
    final method = superclass?.findMethod(expr.method.lexeme);

    if (method == null) {
      throw RuntimeError(
          expr.method, "Undefined property '${expr.method.lexeme}'.");
    }

    return method.bind(obj);
  }

  @override
  Object visitThisExpr(ThisExpr expr) => _lookup(expr.keyword, expr);

  @override
  Object visitUnaryExpr(UnaryExpr expr) {
    final right = _evaluate(expr.right);

    switch (expr.operator.type) {
      case TokenType.MINUS:
        _checkNumberOperand(expr.operator, right);
        return -(right as double);
      case TokenType.BANG:
        return !isTruthy(right);
      default:
        throw Exception('Unreachable');
    }
  }

  @override
  Object visitVariableExpr(VariableExpr expr) => _lookup(expr.name, expr);

  @override
  void visitVarStmt(VarStmt stmt) {
    var value;

    if (stmt.initializer != null) {
      value = _evaluate(stmt.initializer!);
    }

    _env.define(stmt.name.lexeme, value);
  }

  void _execute(Stmt stmt) => stmt.accept(this);

  void executeBlock(List<Stmt> statements, Environment env) {
    final prev = _env;

    try {
      _env = env;
      statements.forEach(_execute);
    } finally {
      _env = prev;
    }
  }

  @override
  void visitWhileStmt(WhileStmt stmt) {
    while (isTruthy(_evaluate(stmt.condition))) {
      _execute(stmt.body);
    }
  }

  Object _evaluate(Expr expr) => expr.accept(this);

  void _checkNumberOperand(Token operator, Object operand) {
    if (operand is! double) {
      throw RuntimeError(operator, 'Operand must be a number.');
    }
  }

  void _checkNumberOperands(Token operator, Object left, Object right) {
    if (!(left is double && right is double)) {
      throw RuntimeError(operator, 'Operands must be numbers.');
    }
  }

  Object _lookup(Token name, Expr expr) {
    final distance = _locals[expr];

    if (distance != null) {
      return _env.getAt(distance, name.lexeme) ?? Nil();
    } else {
      return globals.get(name);
    }
  }
}

File: rlox/src/interpreter.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::mem;
use std::rc::Rc;

use crate::ast::{BinOp, Expr, Keyword, Lit, Stmt, UnOp};
use crate::clock::Clock;
use crate::env::Env;
use crate::function::Func;
use crate::value::Value;
use crate::visitor::{ExprVisitor, StmtVisitor};
use crate::LoxError;

pub struct Interpreter {
    env: Rc<RefCell<Env>>,
    globals: Rc<RefCell<Env>>,
    locals: HashMap<String, usize>,
}

impl Interpreter {
    pub fn new() -> Interpreter {
        let mut env = Env::new();
        env.define("clock", Value::Call(Rc::new(Clock)));
        let globals = Rc::new(RefCell::new(env));

        Interpreter {
            env: globals.clone(),
            globals,
            locals: HashMap::new(),
        }
    }

    pub fn globals(&self) -> Rc<RefCell<Env>> {
        self.globals.clone()
    }

    pub fn interpret(&mut self, stmts: Vec<Stmt>) -> Result<(), LoxError> {
        for stmt in stmts {
            self.execute(&stmt)?
        }

        Ok(())
    }

    pub fn resolve(&mut self, name: &str, depth: usize) {
        self.locals.insert(name.into(), depth);
    }

    pub fn execute_block(&mut self, stmts: &[Stmt], env: Rc<RefCell<Env>>) -> Result<(), LoxError> {
        let prev = mem::replace(&mut self.env, env);
        for stmt in stmts {
            if let Err(e) = self.execute(stmt) {
                self.env = prev;
                return Err(e);
            }
        }

        self.env = prev;
        Ok(())
    }

    fn evaluate(&mut self, expr: &Expr) -> Result<Value, LoxError> {
        self.visit_expr(expr)
    }

    fn execute(&mut self, stmt: &Stmt) -> Result<(), LoxError> {
        self.visit_stmt(stmt)
    }

    #[allow(unused)]
    fn lookup_variable(&mut self, name: &str, line: u64) -> Result<Value, LoxError> {
        let res = if let Some(n) = self.locals.get(name) {
            self.env.borrow_mut().get_at(*n, name)
        } else {
            self.globals.borrow_mut().get(name)
        };

        res.map_err(|e| LoxError::Runtime(e.to_string(), line))
    }
}

impl ExprVisitor for Interpreter {
    type Output = Result<Value, LoxError>;

    fn visit_assign_expr(&mut self, name: &str, expr: &Expr, line: u64) -> Result<Value, LoxError> {
        let val = self.evaluate(expr)?;
        self.env
            .borrow_mut()
            .assign(name, val.clone())
            .map_err(|e| LoxError::Runtime(e, line))?;

        Ok(val)
    }

    fn visit_binary_expr(
        &mut self,
        lhs: &Expr,
        op: BinOp,
        rhs: &Expr,
        line: u64,
    ) -> Result<Value, LoxError> {
        let lhs = self.evaluate(lhs)?;
        let rhs = self.evaluate(rhs)?;

        let res = match op {
            BinOp::Sub => lhs.sub(&rhs),
            BinOp::Div => lhs.div(&rhs),
            BinOp::Mul => lhs.mul(&rhs),
            BinOp::Add => lhs.add(&rhs),
            BinOp::Gt => lhs.gt(&rhs),
            BinOp::GtEq => lhs.ge(&rhs),
            BinOp::Lt => lhs.lt(&rhs),
            BinOp::LtEq => lhs.le(&rhs),
            BinOp::Eq => Ok(Value::from(lhs.eq(&rhs))),
            BinOp::NotEq => Ok(Value::from(lhs.ne(&rhs))),
        };

        res.map_err(|e| LoxError::Runtime(e.to_string(), line))
    }

    fn visit_call_expr(
        &mut self,
        callee: &Expr,
        args: &[Expr],
        line: u64,
    ) -> Result<Value, LoxError> {
        if let Value::Call(fun) = self.evaluate(callee)? {
            let mut values = vec![];
            for arg in args {
                values.push(self.evaluate(arg)?)
            }

            if values.len() != fun.arity() {
                let msg = format!(
                    "Expected {} arguments but got {}.",
                    fun.arity(),
                    values.len()
                );
                return Err(LoxError::Runtime(msg, line));
            }

            fun.call(self, values)
        } else {
            let msg = "Can only call functions and classes.";
            Err(LoxError::Runtime(msg.into(), line))
        }
    }

    fn visit_literal_expr(&mut self, literal: &Lit) -> Self::Output {
        Ok(Value::from(literal))
    }

    fn visit_logical_expr(
        &mut self,
        lhs: &Expr,
        kw: Keyword,
        rhs: &Expr,
        _line: u64,
    ) -> Result<Value, LoxError> {
        let lhs = self.evaluate(lhs)?;

        match kw {
            Keyword::Or if lhs.is_truthy() => Ok(lhs),
            Keyword::And if !lhs.is_truthy() => Ok(lhs),
            _ => self.evaluate(rhs),
        }
    }

    fn visit_unary_expr(&mut self, op: UnOp, rhs: &Expr, line: u64) -> Result<Value, LoxError> {
        let val = self.evaluate(rhs)?;

        let res = match op {
            UnOp::Neg => val.neg(),
            UnOp::Not => Ok(Value::from(!val.is_truthy())),
        };

        res.map_err(|e| LoxError::Runtime(e.to_string(), line))
    }

    fn visit_variable_expr(&mut self, name: &str, line: u64) -> Self::Output {
        self.env
            .borrow_mut()
            .get(name)
            .map_err(|e| LoxError::Runtime(e, line))
    }
}

impl StmtVisitor for Interpreter {
    type Output = Result<(), LoxError>;

    fn visit_block_stmt(&mut self, stmts: &[Stmt]) -> Self::Output {
        self.execute_block(
            &stmts,
            Rc::new(RefCell::new(Env::with_environment(self.env.clone()))),
        )
    }

    fn visit_expression_stmt(&mut self, expr: &Expr) -> Self::Output {
        self.evaluate(expr).map(|_| ())
    }

    fn visit_function_stmt(
        &mut self,
        name: &str,
        params: &[String],
        body: &[Stmt],
        _line: u64,
    ) -> Self::Output {
        let fun = Func::new(name, params, body, self.env.clone());

        self.env
            .borrow_mut()
            .define(name, Value::Call(Rc::new(fun)));

        Ok(())
    }

    fn visit_if_stmt(
        &mut self,
        condition: &Expr,
        then: &Stmt,
        r#else: Option<&Stmt>,
    ) -> Self::Output {
        if self.evaluate(condition)?.is_truthy() {
            self.execute(then)?
        } else if let Some(stmt) = r#else {
            self.execute(stmt)?
        }

        Ok(())
    }

    fn visit_print_stmt(&mut self, expr: &Expr) -> Self::Output {
        self.evaluate(expr).map(|val| println!("{}", val))
    }

    fn visit_return_stmt(&mut self, expr: Option<&Expr>, _line: u64) -> Self::Output {
        let val = match expr {
            Some(exp) => self.evaluate(exp)?,
            None => Value::Nil,
        };

        Err(LoxError::Return(val))
    }

    fn visit_var_stmt(&mut self, name: &str, initializer: Option<&Expr>) -> Self::Output {
        let val = if let Some(v) = initializer {
            self.evaluate(&v)?
        } else {
            Value::Nil
        };

        self.env.borrow_mut().define(&name, val);

        Ok(())
    }

    fn visit_while_stmt(&mut self, condition: &Expr, body: &Stmt) -> Self::Output {
        while self.evaluate(condition)?.is_truthy() {
            self.execute(body)?
        }

        Ok(())
    }
}

impl Default for Interpreter {
    fn default() -> Self {
        Self::new()
    }
}
